From 3932ed794ce97e5b0398bbbe234745b9541dfea2 Mon, 10 Oct 2016 16:16:13 +0200
From: Martin Grimme <martin.grimme@gmail.com>
Date: Tue, 14 Jul 2015 19:03:04 +0200
Subject: [PATCH] [zypp] Fixed crash of packagekitd if zypp is locked by another process during cache switching. Fixes JB#30760


diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index bc64cd2..e8f0dd4 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -190,8 +190,13 @@
 			     const char *format, ...);
 
 static void
-zypp_reset_pool ()
+zypp_reset_pool () // may throw a ZYppFactoryException
 {
+	// get the ZYpp instance first, because it may raise an exception if locked
+	// by another process, and we don't want to modify the pool in that case
+	ZYpp::Ptr zypp = NULL;
+	zypp = ZYppFactory::instance ().getZYpp ();
+
 	RepoManager manager;
 
 	// iterate over all known repositories and reload them from the cache on disk
@@ -215,8 +220,6 @@
 	}
 
 	// reset the state of the resolver
-	ZYpp::Ptr zypp = NULL;
-	zypp = ZYppFactory::instance ().getZYpp ();
 	if (zypp) {
 		zypp->pool().resolver().reset();
 		// the upgrade mode is not reset when resetting the solver and must be
@@ -279,7 +282,13 @@
 
 	// if the cache path was swapped, the pool holds wrong information now and
 	// has to be reset
-	zypp_reset_pool();
+	try {
+		zypp_reset_pool();
+	} catch (const Exception &ex) {
+		PK_ZYPP_LOG("Failed to reset pool: %s",
+			ex.asUserString().c_str());
+		return FALSE;
+	}
 
 	return TRUE;
 }
